src/model/ValidatorFactory.js

```diff
@@ -9,69 +9,110 @@ import FormUtils from "../util/FormUtils";
-        delete result[Constants.HAS_VALIDATION_MESSAGE];
+    };
+  }
+
+  static _isQuestionAnswered(val, question) {
+    return (
+      val !== null &&
+      val !== undefined &&
+      val !== "" &&
+      (!FormUtils.isCheckbox(question) || val !== false)
+    );
+  }
+
+  static _validateAnswer(question, intl, val, validators) {
+    const result = {};
+    for (const validator of validators) {
+      const validationResult = validator(question, intl, val);
+      if (!validationResult.isValid) {
+        result[Constants.HAS_VALID_ANSWER] = false;
+        result[Constants.HAS_VALIDATION_MESSAGE] = validationResult.message;
         return result;
-      };
+      }
     }
+    result[Constants.HAS_VALID_ANSWER] = true;
+    return result;
   }
 
-  static _generateRequiresAnswerValidator(question, intl) {
-    return (answer) => {
-      let val = null;
-      if (answer[Constants.HAS_DATA_VALUE]) {
-        val = JsonLdUtils.getJsonAttValue(answer, Constants.HAS_DATA_VALUE);
-      } else if (answer[Constants.HAS_OBJECT_VALUE]) {
-        val = JsonLdUtils.getJsonAttValue(
-          answer,
-          Constants.HAS_OBJECT_VALUE,
-          "@id"
-        );
+  static patternValidator(question, intl, val) {
+    if (question[Constants.PATTERN]) {
+      let pattern = question[Constants.PATTERN];
+      const regExp = new RegExp(pattern);
+      const isValid =
+        regExp.test(val) ||
+        !ValidatorFactory._isQuestionAnswered(val, question);
+      if (!isValid) {
+        return {
+          isValid: false,
+          message:
+            "Please enter a valid answer to " +
+            JsonLdUtils.getLocalized(question[JsonLdUtils.RDFS_LABEL], intl),
+        };
       }
-      const isValid = val !== null && val !== undefined && val !== "";
-      const result = {};
-      result[Constants.HAS_VALID_ANSWER] = isValid;
-      result[Constants.HAS_VALIDATION_MESSAGE] = isValid
-        ? null
-        : JsonLdUtils.getLocalized(question[JsonLdUtils.RDFS_LABEL], intl) +
-          " is missing a value.";
-      return result;
-    };
+    }
+    return { isValid: true };
   }
 
-  static _generateRequiresAnswerCheckBoxValidator(question, intl) {
-    return (answer) => {
-      let val = null;
-      if (answer[Constants.HAS_DATA_VALUE]) {
-        val = JsonLdUtils.getJsonAttValue(answer, Constants.HAS_DATA_VALUE);
-      } else if (answer[Constants.HAS_OBJECT_VALUE]) {
-        val = JsonLdUtils.getJsonAttValue(
-          answer,
-          Constants.HAS_OBJECT_VALUE,
-          "@id"
-        );
+  static requiredValidator(question, intl, val) {
+    if (question[Constants.REQUIRES_ANSWER]) {
+      const isValid = ValidatorFactory._isQuestionAnswered(val, question);
+      if (!isValid) {
+        return {
+          isValid: false,
+          message:
+            JsonLdUtils.getLocalized(question[JsonLdUtils.RDFS_LABEL], intl) +
+            (FormUtils.isCheckbox(question)
```

Comment:
I believe the checkbox validator should be a separate validator as it was in previous implementation.

Within the list of validators "checkbox validator" should be before "required validator".

Reply:
I am not sure about this, why should the validator care about the type of question? In this case, we need a different validator for each question type, and each validator type with e.g. :
- `checkboxAnswerRequiredValidator`
- `checkboxAnswerCompletenessValidator`
- `regularInputAnswerRequiredValidator`
- `regularInputAnswerCompletenessValidator`
- etc.

Reply:
Ok I understand better